### DISCUSSION TITLE
Is this repository the right place to post non-Blazor issuess/questions re: WASM compilation?

### DISCUSSION THREAD

**Discussion by petertiedemann** [dotnet/runtime#70632] [Q&A]
I have a goal of exposing an existing .NET library as a WASM "library" that can be consumed by existing Javascript applications (so no blazor). I have found it very hard to find out what the "recommended" path is (just finding out which repo contains the WASM compiler bits is pretty tricky, as is determining what its actually called (is it still called mono-wasm?)).

My question is simply whether this is the right place to ask questions about this area? And if not, any pointers to the right place would be greatly appreciated. 

I am currently using Uno.Wasm.Bootstrap, but it does not really seem the right place to ask, as (in my understanding) it is simply a packaging of the existing tooling. Therefore I am wondering if a an issue like the one below, is better raised here?

https://github.com/unoplatform/Uno.Wasm.Bootstrap/issues/515

**Comment by danmoseley** [comment]
@lewing @danroth27

**Comment by danroth27** [comment]
@petertiedemann Yup, this repo is the right place. The .NET WebAssembly build tools workload and the .NET WebAssembly runtime are managed from this repo.

**Comment by lewing** [comment]
Yes, this is the right place.  Feel free to open issues with the problems you are running into and we will take a look at them.

cc @maraf

**Comment by petertiedemann** [comment]
@danroth27 @lewing Thanks for letting me know. Having worked around the specific issue I linked to above (by using a different) setup, I would like to just describe my experience in trying to achieve my goal, as I think it points to some problems (at the very least with documentation). Hope its okay to place here in a discussion (seems too vague for for a specific issue).

As mentioned, I wanted to to create a WASM library wrapping a C# API to be consumed in browser JS. Almost all material on C# and WASM is focused on end-to-end Blazor. I failed completely to find any documentation or guidelines on generating WASM without Blazor. I went through the following attempts (all using latest .NET 7 Preview and VS2022 Preview):

1. Uno.Wasm.Bootstrap : Initial setup worked, but mixed mode AOT failed (as per above) issue.
2. WebAssembly Browser App template: Has no readme/instructions, Running with F5 gives an error about hostpolicy.dll being missing. Contains an index.html and main.js that are not included when publishing. If copied manually to the publish directory and served, shows `The requested module './dotnet.js' does not provide an export named 'default'` in the developer console.
3. WebAssembly Console App template ( I know this is not for browser ) README.md calls for running main.cjs which is not copied when publishing, and if copied manually and run with node complains about missing mono-config.json.
4. Manual setup based on a single project "lean blazor" setup from this [SO answer](https://stackoverflow.com/questions/69743254/how-to-use-c-sharp-webassemly-from-javascript-without-blazor-web-components) : **The winner**

4 turned out great in my case (it actually allowed me to demo our fairly complex library running in the browser. However, I cannot help wondering if I am dragging in unnecessary dependencies here for example.

I guess my questions are:

1. Did I just pick a poor time to look at this? Is the above confusion due to improvements coming ongoing .NET 7.0 work? 
2. Is (or will there be) any "best practice" on how to create libraries for existing JS applications? It seems like it could be a more popular use-case than Blazor for UI considering just how much JS code is already out there.

**Comment by maraf** [comment]
Hi @petertiedemann,

Thanks for sharing your experience. I have created an issue for missing template readme (https://github.com/dotnet/runtime/issues/71047)

There are quite some features important for running .NET on WASM without Blazor that are coming
- https://github.com/dotnet/runtime/pull/68696 - the ability to `dotnet run` on WASM app.
- https://github.com/dotnet/runtime/pull/66304 - interop with JavaScript.
- https://github.com/dotnet/runtime/pull/70746 - switch to ES6.

I have tested the browser template now, with preview5, and after publishing it (`dotnet publish`), I was able run a static file server from an AppBundle directory (`.\bin\Debug\net7.0\browser-wasm\AppBundle`) and see the .NET output in the console. The important thing is to use the "AppBundle" directory, not "publish" (this also applies to the console template).

You should be able to `dotnet run` too, but there was a bug on Windows which is addressed in the next preview.

For your scenario Blazor really seems unnecessary, but without it, you won't be able to interact with the JavaScript, as we don't have a public interop API on the runtime side yet.
